pkg/ccl/backupccl/alter_backup_planning.go

```diff
@@ -0,0 +1,265 @@
+	}
+
+	subdirFn := func() (string, error) { return "", nil }
+	if alterBackupStmt.Subdir != nil {
+		subdirFn, err = p.TypeAsString(ctx, alterBackupStmt.Subdir, "ALTER BACKUP")
+		if err != nil {
+			return nil, nil, nil, false, err
+		}
+	}
+
+	var newKmsFn func() ([]string, error)
+	newKmsFn, err = p.TypeAsStringArray(ctx, tree.Exprs(alterBackupStmt.NewKMSURI), "ALTER BACKUP")
+	if err != nil {
+		return nil, nil, nil, false, err
+	}
+
+	var oldKmsFn func() ([]string, error)
+	oldKmsFn, err = p.TypeAsStringArray(ctx, tree.Exprs(alterBackupStmt.OldKMSURI), "ALTER BACKUP")
+	if err != nil {
+		return nil, nil, nil, false, err
+	}
+
+	fn := func(ctx context.Context, _ []sql.PlanNode, resultsCh chan<- tree.Datums) error {
+		backup, err := fromFn()
+		if err != nil {
+			return err
+		}
+
+		subdir, err := subdirFn()
+		if err != nil {
+			return err
+		}
+
+		if subdir != "" {
+			if strings.EqualFold(subdir, "LATEST") {
+				// set subdir to content of latest file
+				latest, err := readLatestFile(ctx, backup[0], p.ExecCfg().DistSQLSrv.ExternalStorageFromURI, p.User())
+				if err != nil {
+					return err
+				}
+				subdir = latest
+			}
+			if len(backup) != 1 {
+				return errors.Errorf("ALTER BACKUP ... IN can only by used against a single collection path (per-locality)")
+			}
+
+			appendPaths := func(uris []string, tailDir string) error {
+				for i, uri := range uris {
+					parsed, err := url.Parse(uri)
+					if err != nil {
+						return err
+					}
+					parsed.Path = path.Join(parsed.Path, tailDir)
+					uris[i] = parsed.String()
+				}
+				return nil
+			}
+
+			if err = appendPaths(backup, subdir); err != nil {
+				return err
+			}
+		}
+
+		var newKms []string
+		newKms, err = newKmsFn()
+		if err != nil {
+			return err
+		}
+
+		var oldKms []string
+		oldKms, err = oldKmsFn()
+		if err != nil {
+			return err
+		}
+
+		return doAlterBackupPlan(ctx, alterBackupStmt, p, backup, newKms, oldKms)
+	}
+
+	return fn, nil, nil, false, nil
+}
+
+func doAlterBackupPlan(
+	ctx context.Context,
+	alterBackupStmt *tree.AlterBackup,
+	p sql.PlanHookState,
+	backup []string,
+	newKms []string,
+	oldKms []string,
+) error {
+	if len(backup) < 1 || len(backup[0]) < 1 {
+		return errors.New("invalid base backup specified")
+	}
+
+	baseStores, err := p.ExecCfg().DistSQLSrv.ExternalStorageFromURI(ctx, backup[0], p.User())
+	if err != nil {
+		return errors.Wrapf(err, "failed to open backup storage location")
+	}
+	defer baseStores.Close()
+
+	opts, err := readEncryptionOptions(ctx, baseStores)
+	if err != nil {
+		return err
+	}
+
+	ioConf := baseStores.ExternalIOConf()
+
+	// Check that at least one of the old keys works on the backup.
+	// Use the first one that works to decrypt the ENCRYPTION-INFO file(s).
+	var defaultKMSInfo *jobspb.BackupEncryptionOptions_KMSInfo
+	oldKMSFound := false
+	for i := len(oldKms) - 1; i >= 0; i-- {
+		for _, encFile := range opts {
+			defaultKMSInfo, err = validateKMSURIsAgainstFullBackup([]string{oldKms[i]},
```

Comment:
nifty, I like it. Just for my edification, we can't pass the entire oldKMS slice as the first param here cause it's possible it has URIs that weren't used to originally encrypt this backup, and we don't want to fail in that case? In other words oldKMS does not need to be a strict subset of the keys used to encrypt this backup?

Reply:
Yes and I believe that is what the CMEK team wanted, to be able to just give many keys at once in case one was down or the wrong one.